### PR TITLE
test: Handle errors relate network in port-fowarding

### PR DESCRIPTION
**Pull request by utam0k**
## Description

test: Handle errors relate network in port-fowarding

## Related Issue(s)

Relates https://github.com/gitpod-io/gitpod/issues/12248

## How to test

Run the test

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

File: test/pkg/integration/integration.go [hunk 1]
```diff
@@ -57,25 +65,33 @@ func NewPodExec(config rest.Config, clientset *kubernetes.Clientset) *PodExec {
...
 		}
 		break
 	}
 	return in, out, errOut, nil
 }
 
+func shouldRetry(count int, err error) bool {
+	if count < connectFailureMaxTries {
+		return err.Error() == errorDialingBackendEOF
+	}
```
Comment (jenting) [inline test/pkg/integration/integration.go:91]: Should we check if the error is errorDialingBackendEOF, then retry?

File: test/pkg/integration/integration.go [hunk 2]
```diff
@@ -57,25 +65,33 @@ func NewPodExec(config rest.Config, clientset *kubernetes.Clientset) *PodExec {
...
 		copyOptions.Container = containername
 		err := copyOptions.Run([]string{src, dst})
 		if err != nil {
-			if !errors.Is(err, io.EOF) {
-				return nil, nil, nil, fmt.Errorf("Could not run copy operation: %v", err)
+			if !shouldRetry(count, err) {
+				return nil, nil, nil, fmt.Errorf("could not run copy operation: %v", err)
 			}
 			time.Sleep(10 * time.Second)
+			count = count + 1
```
Comment (jenting) [inline test/pkg/integration/integration.go:80]: We have count++ at line 68, is this line still needed?

**Review by jenting — COMMENTED** [review]
Thanks for the PR.
But from my testing recent days, I did not encounter the port-forward error.
So I think should we need to enhance the port-forward error again because you have enhanced it in the last PR already.

**Comment by utam0k** [comment]
> Thanks for the PR.
> But from my testing recent days, I did not encounter the port-forward error.
> So I think should we need to enhance the port-forward error again because you have enhanced it in the last PR already.

Have you ran all tests from werft?

File: test/pkg/integration/integration.go (hunk 2 above)
Comment (utam0k) [reply test/pkg/integration/integration.go:80]: you are right!

File: test/pkg/integration/integration.go (hunk 1 above)
Comment (utam0k) [reply test/pkg/integration/integration.go:91]: Yes.

**Comment by jenting** [comment]
> > Thanks for the PR.
> > But from my testing recent days, I did not encounter the port-forward error.
> > So I think should we need to enhance the port-forward error again because you have enhanced it in the last PR already.
> 
> Have you ran all tests from werft?

Yes, about five times.

**Comment by utam0k** [comment]
> > > Thanks for the PR.
> > > But from my testing recent days, I did not encounter the port-forward error.
> > > So I think should we need to enhance the port-forward error again because you have enhanced it in the last PR already.
> > 
> > 
> > Have you ran all tests from werft?
> 
> Yes, about five times.

@jenting 
This appears to be an error in port-forwarding. May I ask you to tell me what your job is?
https://werft.gitpod-dev.com/job/gitpod-workspace-run-integration-tests-main.118/raw

**Comment by jenting** [comment]
> > > > Thanks for the PR.
> > > > But from my testing recent days, I did not encounter the port-forward error.
> > > > So I think should we need to enhance the port-forward error again because you have enhanced it in the last PR already.
> > > 
> > > 
> > > Have you ran all tests from werft?
> > 
> > 
> > Yes, about five times.
> 
> @jenting This appears to be an error in port-forwarding. May I ask you to tell me what your job is? https://werft.gitpod-dev.com/job/gitpod-workspace-run-integration-tests-main.118/raw

🤔 I might be wrong. I did not check the lastest workspace integration tests result.

**Comment by utam0k** [comment]
@jenting At least, around these codes often get failed when I run it from my local
https://github.com/gitpod-io/gitpod/blob/ded3a97e9e858c411ea060c7deade206d2f38a44/test/pkg/integration/integration.go#L95-L97
```console
--- FAIL: TestCreateBucket (5.78s)
    --- FAIL: TestCreateBucket/DaemonAgent.CreateBucket (5.78s)
        --- FAIL: TestCreateBucket/DaemonAgent.CreateBucket/it_should_create_a_bucket (5.78s)
            storage_test.go:32: Could not run copy operation: error dialing backend: EOF
```

File: test/pkg/integration/integration.go [hunk 3]
```diff
@@ -187,6 +203,43 @@ func Instrument(component ComponentType, agentName string, namespace string, kub
 		return nil, closer, err
 	}
 
+	var (
+		res *rpc.Client
+		cl  []func() error
+	)
+	for i := 0; i < 10; i++ {
```
Comment (jenting) [inline test/pkg/integration/integration.go:210]: Should we use connectFailureMaxTries instead of 10?

Comment (utam0k) [reply test/pkg/integration/integration.go:210]: Great indicate

**Comment by utam0k** [comment]
@jenting Thanks for your review. PTAL

File: test/pkg/integration/integration.go (hunk 1 above)
Comment (jenting) [reply test/pkg/integration/integration.go:90]: I haven't seen you change this. 🤔

Comment (utam0k) [reply test/pkg/integration/integration.go:90]: @jenting 
Sorry, I may misunderstand. Can I ask you to rephrase?

Comment (jenting) [reply test/pkg/integration/integration.go:90]: Never mind, my thought was the same logic
```go
        if err.Error() == errorDialingBackendEOF && count < connectFailureMaxTries {
		return true
	}
        return false
```

**Review by jenting — APPROVED** [review]

**Pull request merged**